src/cljc/athens/common_events/resolver.cljc

```diff
@@ -861,6 +861,47 @@
     tx-data))
 
 
+(defmethod resolve-event-to-tx :datascript/selected-delete
+  [db {:event/keys [args]}]
+  ;; We know that we only need to dec indices after the last block. The former blocks
+  ;; are necessarily going to remove all tail children, meaning we only need to be
+  ;; concerned with the last N blocks that are selected, adjacent siblings, to
+  ;; determine the minus-after value.
+  (println "resolver :datascript/drop-multi-child args" (pr-str args))
+  (let [{:keys [uids]}                    args
+        selected-sibs-of-last             (->> uids
+                                               (d/q '[:find ?sib-uid ?o
+                                                      :in $ ?uid [?selected ...]
+                                                      :where
+                                                      ;; get all siblings of the last block
+                                                      [?e :block/uid ?uid]
+                                                      [?p :block/children ?e]
+                                                      [?p :block/children ?sib]
+                                                      [?sib :block/uid ?sib-uid]
+                                                      ;; filter selected
+                                                      [(= ?sib-uid ?selected)]
+                                                      [?sib :block/order ?o]]
+                                                    db
+                                                    (last uids))
+                                               (sort-by second))
+        [uid order]                       (last selected-sibs-of-last)
+        parent-eid                        (:db/id (common-db/get-parent db [:block/uid uid]))
+        n                                 (count selected-sibs-of-last)
+        reindex                           (common-db/minus-after db
+                                                                 parent-eid
+                                                                 order
+                                                                 n)
+        retracted-vec                     (mapv (fn [uid]
```

Comment:
Yeah you are right, `retract-uid-recursively` uses `:db/retractEntity`, while `:db/retract` as you said only removes ref between parent and child.

From the user's point of view if a block is deleted while it is referenced somewhere else, after deletion user will see  `((some random block uid))` instead of the referenced block content (which is now deleted) and it is right implementation. But personally, I would not like to see `((some random block uid))` when the original block is deleted, maybe in some cases I would like to delete the original block but would like to keep its content referenced somewhere else.

What does "delete" mean actually? It is not like this data once deleted cannot be recovered, and if by delete means removing all the data (including block refs, relationships, block embeds etc) related to this block to be removed then we fail slightly short by not removing blocks refs and embeds (which may be a lot of computation). What do you think?

Reply:
Alex and I discussed that case a while back. The real question being "what happens to refs in text when the ref'd thing is deleted". It applies to anything that can be reffed (pages and blocks right now).

The decision we came to is that the refs should be replaced with their content, and that it's the events responsibility to do so, which will then trigger linkmaker to remove the `:block/refs`. The idea is that the strings are the source of truth and now they changed.

So for instance if you have `here is ((block-uid))` and `block-uid` has `an orange` as `:block/string`, deleting `block-uid` would leave the other block with `here is an orange` as `:block/string`. If it was a title instead, the same would happen.

I'm not sure this happens right now. Opened https://github.com/athensresearch/athens/issues/1448 to track it, and added it to the project board for alpha-1.

Reply:
To answer your other point ("It is not like this data once deleted cannot be recovered"), I don't think data recovery needs to be complected with deletion semantics.

It's possible to do recovery by others means (backups, time travel, etc), and deletion semantics can impose different restrictions (e.g. full deletion as per GDPR, link consistency, removal from search results, maintaining a minimal graph, etc). 

It's possible these two interact with each other but not necessary. In particular we shouldn't bake in assumptions of how we expect data recovery to work later on.